### DISCUSSION TITLE
TransparentProxy deploy problem

### DISCUSSION THREAD

**Discussion by gethub2022** [zkSync-Community-Hub/zksync-developers#57] [Deployments] [answered]
### Environment

Testnet

### zkSolc Version

latest

### zksync-web3 Version

0.14.3

### Hardhat.config.ts

{
  "name": "hardhat-project",
  "dependencies": {
    "@ethersproject/providers": "^5.7.2",
    "@nomicfoundation/hardhat-chai-matchers": "^2.0.1",
    "@nomicfoundation/hardhat-network-helpers": "^1.0.8",
    "@nomicfoundation/hardhat-toolbox": "^2.0.2",
    "@nomiclabs/hardhat-etherscan": "^3.1.7",
    "@openzeppelin/contracts": "^4.8.2",
    "@openzeppelin/contracts-upgradeable": "^4.8.2",
    "@openzeppelin/hardhat-upgrades": "^1.22.1",
    "@openzeppelin/upgrades": "^2.8.0",
    "@typechain/ethers-v5": "^10.2.0",
    "@typechain/hardhat": "^6.1.5",
    "@types/chai": "^4.3.4",
    "@types/mocha": "^10.0.1",
    "@types/node": "^20.4.0",
    "chai": "^4.3.7",
    "dotenv": "^16.0.3",
    "ethers": "^5.7.2",
    "hardhat": "^2.16.1",
    "hardhat-deploy": "^0.11.25",
    "hardhat-deploy-ethers": "^0.3.0-beta.13",
    "hardhat-gas-reporter": "^1.0.9",
    "moment": "^2.29.4",
    "solidity-coverage": "^0.8.2",
    "solmate": "^6.7.0",
    "ts-node": "^10.9.1",
    "typechain": "^8.1.1",
    "typescript": "^5.1.6",
    "web3-utils": "^1.8.2"
  },
  "version": "1.0.0",
  "main": "index.js",
  "repository": "aaa",
  "author": "caglarusta <t.caglar.usta@gmail.com>",
  "license": "MIT",
  "devDependencies": {
    "@ethersproject/hash": "^5.7.0",
    "@ethersproject/web": "^5.7.1",
    "@matterlabs/hardhat-zksync-deploy": "^0.6.3",
    "@matterlabs/hardhat-zksync-solc": "^0.4.1",
    "@nomiclabs/hardhat-ethers": "npm:hardhat-deploy-ethers@^0.4.1",
    "zksync-web3": "^0.14.3"
  }
}

### Deployment Script (WITHOUT PRIVATE KEY)

import { HardhatRuntimeEnvironment } from 'hardhat/types';
import { ABI, DeployFunction } from 'hardhat-deploy/types';
import { getCreationCode } from '../utils/utils';
import salts from '../utils/salts';
import { Deployer } from "@matterlabs/hardhat-zksync-deploy";
import { utils, Wallet } from "zksync-web3";
import dotenv from 'dotenv';
dotenv.config()
const PRIVATE_KEY = process.env.PRIVATE_KEY
if (!PRIVATE_KEY) {
  throw Error("process.env.PRIVATE_KEY IS REQUIRED");
}

const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {

    const { deployments, ethers } = hre;
    const wallet = new Wallet(PRIVATE_KEY);
    const deployer = new Deployer(hre,wallet);
    const Bit5Treasury = await ethers.getContractFactory("Bit5Treasury");
    const Create3Factory = await deployments.get("Create3Factory")
    const create3Factory = await ethers.getContractAt("Create3Factory", Create3Factory.address)

    const ProxyAdmin = await deployments.get("ProxyAdmin");
    const proxyAdmin = await ethers.getContractAt("ProxyAdmin", ProxyAdmin.address)

    const bit5Treasury = await Bit5Treasury.deploy();
    await bit5Treasury.deployed();

    const functionData = bit5Treasury.interface.encodeFunctionData("initialize", deployer);

         const TransparentProxyCreationCode = await getCreationCode({
         contractName: "TransparentProxy", // In zksync the strandard proxy deployment with hardhat deploy not works
        constructorArgs: {
            types: ["address", "address", "bytes"],
            values: [bit5Treasury.address, proxyAdmin.address, functionData]
        }
    })
  

    const tx2 = await create3Factory.deploy(salts.Bit5Treasury.salt, TransparentProxyCreationCode);
    const receipt = await tx2.wait();
    const parsedEvents = receipt.events?.map((event: any) => {
        return {
            name: event.event,
            args: event.args
        }
    });

    if (!parsedEvents) {
        throw new Error("No events found");
    }

    const { contractAddress } = parsedEvents.find((event: any) => event.name === "Deployed")?.args

    const treasury = Bit5Treasury.attach(contractAddress);

    console.log("Bit5Treasury deployed to:", treasury.address);

    await deployments.save("Bit5Treasury", {
        address: treasury.address,
        abi: (JSON.parse(Bit5Treasury.interface.format("json") as string) as ABI),
        
        
        
    })

};
export default func;

### Contract Code

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
import "./interfaces/IBit5Treasury.sol";
import {Bit5Errors} from "./interfaces/Errors.sol";

contract Bit5Treasury is
    Initializable,
    PausableUpgradeable,
    OwnableUpgradeable,
    IBit5Treasury,
    Bit5Errors
{
    event Withdraw(
        address indexed fromCollection,
        address indexed tokenAddress,
        uint256 amount
    );
    event WithdrawRaw(address indexed tokenAddress, uint256 amount);
    event Deposit(
        address indexed collectionAddress,
        address indexed tokenAddress,
        uint256 amount
    );

    mapping(address => uint256) public collectionBalances;

    address public bit5;

    /// @custom:oz-upgrades-unsafe-allow constructor
    constructor() {
        _disableInitializers();
    }

    function initialize(address _owner) public initializer {
        __Pausable_init();
        __Ownable_init();
        transferOwnership(_owner);
    }

    function pause() public onlyOwner {
        _pause();
    }

    function unpause() public onlyOwner {
        _unpause();
    }

    function deposit(
        address collectionAddress,
        address tokenAddress,
        uint256 amount
    ) external {
        if (msg.sender != bit5) revert Unauthorized();
        IERC20(tokenAddress).transferFrom(
            bit5,
            address(this),
            amount
        );

        // if (!res) revert TokenTransferFailed();

        collectionBalances[collectionAddress] += amount;

        emit Deposit(collectionAddress, tokenAddress, amount);
    }

    function setBit5(address _bit5) external onlyOwner {
        bit5 = _bit5;
    }

    function withdraw(
        address fromCollection,
        address tokenAddress,
        uint256 amount
    ) external onlyOwner {
        if (amount > collectionBalances[fromCollection]) {
            revert UnsufficientTreasuryFunds();
        }

        collectionBalances[fromCollection] -= amount;

        IERC20(tokenAddress).transfer(msg.sender, amount);

        emit Withdraw(fromCollection, tokenAddress, amount);
    }
}

### Does this work on other EVMs? (If yes, please list at least 1 of them)

Yes

### Description of What Your Contract Does

It helps to deploy on the same proxy address

### Additional Details

const TransparentProxyCreationCode = await getCreationCode({
        contractName: "TransparentProxy", // In zksync the strandard proxy deployment with hardhat deploy not works
        constructorArgs: {
            types: ["address", "address", "bytes"],
            values: [myContract.address, proxyAdmin.address, functionData]
        }
    })
  Return Error => ProviderError: Failed to serialize transaction: toAddressIsNull

  has anyone encountered this problem and what can we do about the solution?

**Comment by uF4No** [comment, marked as answer]
Looks like you're not using the [`hardhat-zksync-upgradable` plugin](https://www.npmjs.com/package/@matterlabs/hardhat-zksync-upgradable). You can find the [example to deploy transparent proxies here](https://era.zksync.io/docs/tools/hardhat/hardhat-zksync-upgradable.html#transparent-upgradable-proxies)

**Comment by martinerinv** [comment]
It seems like the issue you're encountering is related to serialization error when deploying a contract through a transparent proxy on zkSync. The error message "toAddressIsNull" suggests that there might be an issue with the target address of the transaction being null.

One potential solution is to ensure that the target address for the transaction is properly set when deploying the transparent proxy. Make sure that the myContract.address variable contains the correct address of the contract you want to proxy. Additionally, verify that the functionData contains the encoded function call data for the initialization of the contract.

If you're still facing issues, double-check the documentation and examples provided by zkSync and the SDK you're using for any specific requirements or considerations when deploying contracts on zkSync. Additionally, consider reaching out to the community or the maintainers of the SDK for further assistance and guidance.

**Comment by evmeiwear** [comment]
Issue: ProviderError: Failed to serialize transaction: toAddressIsNull
Description: Error occurs when deploying a contract using a transparent proxy on zkSync with hardhat-deploy.
Possible Solution: Ensure the toAddress is not null in the transaction serialization process. Verify the contract deployment setup and functionData.

**Comment by mcleodjv** [comment]
It seems like the error "ProviderError: Failed to serialize transaction: toAddressIsNull" is occurring due to an issue with the serialization of the transaction object, specifically with the toAddress field being null. This could be caused by various factors, such as incorrect contract address or invalid transaction data.

To address this issue, you may need to double-check the contract addresses being used, ensure that the functionData passed to the constructor is correct, and verify that all required parameters are properly set. Additionally, review the TransparentProxyCreationCode generation process to ensure that it accurately reflects the constructor arguments expected by the contract.

If the issue persists, consider debugging the transaction serialization process and inspecting the generated transaction object to identify any discrepancies or errors.

**Comment by lashawnheisler** [comment]
The error "ProviderError: Failed to serialize transaction: toAddressIsNull" typically occurs when the contract address passed to the proxy creation code is null or invalid. Check that the myContract.address variable contains a valid contract address before deploying the proxy contract. Additionally, ensure that the constructorArgs object includes the correct types and values required for the contract initialization.
